### PR TITLE
feat(light client): fetch block body from remote

### LINKED ISSUE #1445: Light clients should fetch block body from remote.
but currently only returns `Ok(None)`.

see https://github.com/paritytech/substrate/blob/a2dfec4d3c467a481da70efb34f105e141a289fa/core/client/src/light/blockchain.rs#L145

### PR DESCRIPTION
**Pull request by niklasad1**
Attempt to close #1445

### CODE REVIEW

File: core/client/src/light/fetcher.rs [hunk 1]
```diff
@@ -108,6 +108,15 @@ pub struct ChangesProof<Header: HeaderT> {
 	pub roots_proof: Vec<Vec<u8>>,
 }
 
+/// Remote block body request
+#[derive(Clone, Default, Debug, PartialEq, Eq, Hash)]
+pub struct RemoteBodyRequest<Header: HeaderT> {
+	/// Header of the requested block body
+	pub header: Header,
```
Comment (andresilva) [inline core/client/src/light/fetcher.rs]: We probably only need the hash.

File: core/client/src/light/blockchain.rs [hunk 2]
```diff
@@ -144,9 +144,20 @@ impl<S, F, Block> BlockchainHeaderBackend<Block> for Blockchain<S, F> where Bloc
...
+				return Ok(None)
+			}
+		};
+
+		self.fetcher().upgrade().ok_or(ClientError::NotAvailableOnLightClient)?
+			.remote_body(RemoteBodyRequest {
+				header,
+				retry_count: None,
+			})
+			.into_future().wait()
```
Comment (DemiMarie-parity) [inline core/client/src/light/blockchain.rs]: We should really return a `Future` instead of blocking.

File: core/client/src/light/fetcher.rs [hunk 3]
```diff
@@ -432,6 +446,10 @@ pub mod tests {
 		fn remote_changes(&self, _request: RemoteChangesRequest<Header>) -> Self::RemoteChangesResult {
 			err("Not implemented on test node".into())
 		}
+
+		fn remote_body(&self, _request: RemoteBodyRequest<Header>) -> Self::RemoteBodyResult {
+			err("Not implemented on test node".into())
```
Comment (DemiMarie-parity) [inline core/client/src/light/fetcher.rs]: This should be refactored into a utility function.

File: core/network/src/on_demand.rs [hunk 4]
```diff
@@ -316,6 +324,45 @@ impl<B> OnDemandService<B> for OnDemand<B> where
...
+				// Number of bodies are hardcoded to 1 for valid `RemoteBodyResponses`
+				if num_bodies != 1 {
+					return Accept::CheckFailed("RemoteBodyResponse: invalid number of blocks".into(),
+												RequestData::RemoteBody(request, sender))
+				}
+
+				let response = response
+					.blocks
+					.into_iter()
+					.take(1)
```
Comment (DemiMarie-parity) [inline core/network/src/on_demand.rs]: This `.take(1)` does noting, if I understand correctly.

File: core/client/src/light/fetcher.rs (hunk 1 above)
Comment (svyatonik) [reply core/client/src/light/fetcher.rs]: The header could be pruned (replaced with CHT) when request+response messages are routing from/to light node => to check it we'll need to retrieve header again from the full node. Holding header in the request guarantees that we can verify response without additional messages (we need to check extrinsics root for that).

Comment (svyatonik) [reply core/client/src/light/fetcher.rs]: Oh, I see that @niklasad1 is also fetching header here. So probably you're right

File: core/network/src/on_demand.rs [hunk 5]
```diff
@@ -316,6 +324,45 @@ impl<B> OnDemandService<B> for OnDemand<B> where
...
+					}).nth(0);
+
+				// Body and Header should included in valid `RemoteBodyResponses`
+				let (body, header) = match response {
+					Some(Some((b, h))) => (b, h),
+					_ => return Accept::CheckFailed("RemoteBodyResponse: is missing body or header".into(),
+													RequestData::RemoteBody(request, sender)),
+				};
+
+				if request.header.extrinsics_root() == header.extrinsics_root() {
```
Comment (svyatonik) [inline core/network/src/on_demand.rs]: Think there's a problem with this check - it doesn't actually guarantee that the returned body matches the header we have requested. It only checks that extrinsic root is matched => this way full node could return header the light node expects + any body. Or am I missing something?

I'd personally in favor of fetching only `BODY` + computing extrinsic root using `HashFor::<Block>::ordered_trie_root(body.iter().map(Encode::encode))` + checking it against local `header.extrinsic_root()`. But this will only work if runtime actually puts that value to the `Header::extrisic_root`. Right now iiuc there are no assumptions of that in the native code, except for single `debug_assert_eq` in `BlockBuilder`. So probably someone else have an idea if we could rely on that fact? BTW - this way we could remove extra network messages && treat `BlockResponse` with only body filled as a response for `on_demand` - iirc Andre did the same with justification requests.

File: core/network/src/on_demand.rs (hunk 4 above)
Comment (niklasad1) [reply core/network/src/on_demand.rs]: yes, thanks

File: core/client/src/light/blockchain.rs (hunk 2 above)
Comment (niklasad1) [reply core/client/src/light/blockchain.rs]: Yeah I agree, but this is a part of the `trait Backend` impl which doesn't support futures. I guess this relies on `http timeouts` instead?! 

However, this is out-of-scope of this PR and could probably refactored to be async instead.

File: core/client/src/light/fetcher.rs (hunk 3 above)
Comment (niklasad1) [reply core/client/src/light/fetcher.rs]: Are you implying that all `test helpers` should be utility functions?

Otherwise please explain!

File: core/network/src/on_demand.rs (hunk 5 above)
Comment (niklasad1) [reply core/network/src/on_demand.rs]: Sounds much better to me, but I'm not familiar with the runtime

File: core/client/src/light/fetcher.rs (hunk 3 above)
Comment (DemiMarie-temp) [reply core/client/src/light/fetcher.rs]: (Same person, just different account)

I meant that we could have a

```rust
fn<T> not_implemented_in_tests() -> Result<T, Error> {
	err("Not implemented on test node".into())
}
```

File: core/network/src/on_demand.rs (hunk 5 above)
Comment (niklasad1) [reply core/network/src/on_demand.rs]: @svyatonik based on Gav's response, shall we put this on ice or just compare `request.header.hash() == response.header.hash()` for now?

File: core/network/src/on_demand.rs [hunk 6]
```diff
@@ -391,6 +402,56 @@ impl<B> OnDemandService<B> for OnDemand<B> where
...
+					.nth(0);
+
+				// Body and Header should included in valid `RemoteBodyResponses`
+				let (body, header) = match response {
+					Some(Some((b, h))) => (b, h),
+					_ => return Accept::CheckFailed("RemoteBodyResponse: is missing body or header".into(),
+													RequestData::RemoteBody(request, sender)),
+				};
+
+				// TODO(svyatonik/niklasad1): ideally we should fetch only:
```
Comment (niklasad1) [inline core/network/src/on_demand.rs]: create an issue for this before merge....

File: core/network/src/on_demand.rs (hunk 5 above)
Comment (svyatonik) [reply core/network/src/on_demand.rs]: IMO let's do as I've proposed ^^^ && file issue that current native (light client) code doesn't support situation when runtime returns anything else (i.e. not `ordered_trie_root(extrinsics)`) from `extrinsic_root()`. I see no other way how we could solve this issue. Committing vulnerable code IMO isn't an option.

Comment (DemiMarie-temp) [reply core/network/src/on_demand.rs]: We can solve this at compile-time by requiring `extrinsic_root` to always return `ordered_trie_root(extrinsics)`, unless a Cargo feature is specified that disables light client support.

File: core/network/src/protocol.rs [hunk 7]
```diff
@@ -365,10 +369,15 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
 			GenericMessage::Status(s) => self.on_status_message(who, s),
 			GenericMessage::BlockRequest(r) => self.on_block_request(who, r),
 			GenericMessage::BlockResponse(r) => {
-				if let Some(request) = self.handle_response(who.clone(), &r) {
-					let outcome = self.on_block_response(who.clone(), request, r);
-					self.update_peer_info(&who);
-					return outcome
+				// Note, this is safe because only `ordinary bodies` and `remote bodies` are received in this matter.
+				if self.is_on_demand_response(&who, r.id) {
```
Comment (niklasad1) [inline core/network/src/protocol.rs]: Perform this check first because otherwise the peer might be disconnected/punished.

Comment (niklasad1) [reply core/network/src/protocol.rs]: /cc @andresilva something like you had in mind? 

It is kind of ugly to introduce `is_on_demand_response` but because of `encapsulation/trait object` I don't see any other options!

File: core/client/src/light/blockchain.rs [hunk 8]
```diff
@@ -144,9 +144,20 @@ impl<S, F, Block> BlockchainHeaderBackend<Block> for Blockchain<S, F> where Bloc
 }
 
 impl<S, F, Block> BlockchainBackend<Block> for Blockchain<S, F> where Block: BlockT, S: Storage<Block>, F: Fetcher<Block> {
-	fn body(&self, _id: BlockId<Block>) -> ClientResult<Option<Vec<Block::Extrinsic>>> {
-		// TODO: #1445 fetch from remote node
-		Ok(None)
+	fn body(&self, id: BlockId<Block>) -> ClientResult<Option<Vec<Block::Extrinsic>>> {
+		let header = match self.header(id) {
+			Ok(Some(header)) => header,
+			_ => {
```
Comment (svyatonik) [inline core/client/src/light/blockchain.rs]: Probably return `Err` if `self.header` has also returned an error?

File: core/client/src/light/fetcher.rs [hunk 9]
```diff
@@ -135,6 +144,8 @@ pub trait Fetcher<Block: BlockT>: Send + Sync {
 	type RemoteCallResult: IntoFuture<Item=Vec<u8>, Error=ClientError>;
 	/// Remote changes result future.
 	type RemoteChangesResult: IntoFuture<Item=Vec<(NumberFor<Block>, u32)>, Error=ClientError>;
+	/// Remote block body result future.
+	type RemoteBodyResult: IntoFuture<Item = Option<Vec<Block::Extrinsic>>, Error = ClientError>;
```
Comment (svyatonik) [inline core/client/src/light/fetcher.rs]: I think you should use `Vec<Block::Extrinsic>` instead of `Option<Vec<Block::Extrinsic>>` here - the block is header + body even though body could be empty.

Comment (niklasad1) [reply core/client/src/light/fetcher.rs]: Yeah I agree, I choose `Option<Vec<Block::Extrinsic>>` because of https://github.com/paritytech/substrate/blob/master/core/client/src/blockchain.rs#L78 but changed now.

**Comment by niklasad1** [comment]
> Probably add couple of tests (positive, when response is accepted and negative, when that's extrinsic root mismatch) to client/src/light/fetcher.rs (could be done in follow up PR)?

I will take of that in this PR

File: core/client/src/light/fetcher.rs [hunk 10]
```diff
@@ -191,6 +205,12 @@ pub trait FetchChecker<Block: BlockT>: Send + Sync {
 		request: &RemoteChangesRequest<Block::Header>,
 		proof: ChangesProof<Block::Header>
 	) -> ClientResult<Vec<(NumberFor<Block>, u32)>>;
+	/// Check remote body proof.
+	fn check_body_proof(
```
Comment (niklasad1) [inline core/client/src/light/fetcher.rs]: moved the `extrinsic_root` verification here to test it!
